TODO-006: Edit a task title

fravila08 opened the pull request:
Acceptance Criteria

Given a list of tasks
When an individual task is displayed
Then a "Edit" button appears to the right of the task to the left of the check box

Given an edit button
When the button is clicked
Then a text input replaces the task title
And the current title is automatically filled into the text box
And the edit button replaces the "Confirm" button
And a "Cancel" button replaces the check box

Given no input is provided in the text box
Then the submit button is disabled

Given input is given in the text box
When the submit button is clicked
Then the new title is displayed as text
And the Confirm and Cancel buttons are replaced by the "Edit" button and checkbox

Given the cancel button is clicked
Then the original title is displayed as text
And the Confirm and Cancel buttons are replaced by the "Edit" button and checkbox

https://user-images.githubusercontent.com/105952966/219450850-e7a77631-28e3-4308-9b69-45da54e22be5.mov

nadiacollado commented on to_do_client/src/__tests__unit_/TaskUnit.test.tsx:
```diff
@@ -34,6 +38,17 @@ describe("Task", () => {
   });
 
+  describe("changeTaskTitle()", () => {
+    it("will return if the changing of the title was successful", async () => {
```
will return true here :)

nadiacollado commented on to_do_server/to_do_app/tests/test_urls.py:
```diff
@@ -60,28 +60,41 @@ def test_change_multiple_PUT_func(self):
         self.assertEquals(resolve(url).func.view_class, Multi_task_handler)
         
-    def test_delete_task_proper(self):
+    def test_delete_task_DELETE_proper(self):
```
Hmm, maybe this can simply be "test_task_DELETE_proper"

Kotauror replied:
I'm still not convinced that `proper` is a good naming practice. The test name should explain the assertion in an unambiguous, non-opinionated way, while `proper` doesn't tell us much about the expectation. Same applies to `improper`.

nadiacollado commented on to_do_server/to_do_app/tests/test_urls.py:
```diff
@@ -60,28 +60,41 @@ def test_change_multiple_PUT_func(self):
+        response = self.client.put(reverse("changetitle", args=[1]), {"name": 1}, "application/json")
+        self.assertEquals(response.status_code, 200)
+        
+    def test_change_task_title_fun(self):
```
I think this is a typo: `func`

nadiacollado commented on to_do_server/to_do_app/tests/test_urls.py:
```diff
@@ -60,28 +60,41 @@ def test_change_multiple_PUT_func(self):
         self.assertEquals(resolve(url).func.view_class, Task_handler)   
         
-    def test_delete_mult_task_proper(self):
+    def test_delete_mult_task_DELETE_proper(self):
```
A little inconsistent to how you're titling one of the other tests. Some use `multiple` in the title, this uses `mult`

Kotauror replied:
Avoid abbreviations in general.

nadiacollado commented on to_do_server/to_do_app/tests/test_views.py:
```diff
@@ -82,4 +82,21 @@ def test_delete_multiple_task_PROPER_input(self):
+        response = self.client.put(reverse("changetitle", args=[1]), {"name":"new title"}, "application/json")
+        body = json.loads(response.content)
+        with self.subTest():
+            self.assertNotEquals(Task.objects.get(id = task1.id), "new title")
```
Think you're missing the `.title` here since you're checking the titles aren't equal, so: `Tast.objects.get(id = task1.id).title, "new title")`

nadiacollado commented on to_do_client/src/components/Task.tsx:
```diff
@@ -36,13 +40,27 @@ export const deleteTask = async (id: number) => {
 }) => {
+  const [taskTitle, setTaskTitle] = useState<string>(task.title);
+  const [showForm, setShowForm] = useState<boolean>(true);
+  const [newTitle, setNewTitle] = useState<string>(task.title)
```
Wondering if we need to have separate state for this. Did you editing the titles using just`taskTitle` and `setTaskTitle`?

fravila08 replied:
The reasoning behind having 2 separate useStates her is to haver the ability to access the original title in case the user would like to cancel the new input value.

nadiacollado reviewed (commented):
Mostly minor comments but if you could take a look at the question I left in the `Task` component

Kotauror commented on to_do_client/src/components/Task.tsx:
```diff
@@ -36,13 +40,27 @@ export const deleteTask = async (id: number) => {
 };
 
+export const changeTaskTitle = async (id:number, name:string) => {
+  try {
```
I think that this `try` `catch` might be unnecessary. This might apply to other places in your code, but I didn't think about it earlier, so sorry if this feedback is delayed. 

In your backend logic, in `to_do_server/to_do_app/views.py#26`, you're already catching all possible errors related to it. And if there is an error, you still return a valid response, but with `changed` set to `False`. This means that BE will never return error. So the `try` on line 44 in this file should never catch an error. 

This is a very common design question about how to handle this kind of situation. One could claim - on the FE we'd like to be independent from the BE. We don't know what BE will do, so we try/catch anyway. Or you could argue otherwise - the system should stay consistent and not do catching of the same thing twice. 

I personally prefer not catching the errors on the BE and let them 'bubble up' to the FE as error responses, but there certainly are different approaches to it. 

That being said - I wonder what made you do the error-catching in both places and if you're planning to continue doing it for the future requests.

fravila08 commented on to_do_client/src/components/Header.tsx:
```diff
@@ -130,19 +132,33 @@ export const Header: React.FC<HeaderProps> = ({
       <Row>
         <Col xs={4} style={{ display: "flex" }}>
           <Button
+            style={{
```
This styling seems to be unique and not repetitive. Is it better to create a class solely for one element or leave it within inline style?

jay-meister replied:
Agree with @nadiacollado, the style object is identical in the following two buttons:

```jsx
          <Button
            style={{
              height: "5vh",
              display: "flex",
              justifyContent: "center",
              alignItems: "center",
              width: "10vw",
            }}
            onClick={changingMultipleStatus}
            disabled={isChangeStatusDisabled()}
            id="changeStatusBtn"
          >
            <img src={change} style={{ height: "6vh" }} />
          </Button>
          <Button
            style={{
              height: "5vh",
              display: "flex",
              justifyContent: "center",
              alignItems: "center",
              width: "10vw",
            }}
            variant="danger"
            onClick={deleteMultipleTasks}
            disabled={isChangeStatusDisabled()}
            id="DeleteMultBtn"
          >
            <img src={trash} style={{ height: "3vh" }} />
          </Button>
```


```

```jsx
          <Button
            style={buttonStyles}
            onClick={changingMultipleStatus}
            disabled={isChangeStatusDisabled()}
            id="changeStatusBtn"
          >
            <img src={change} style={{ height: "6vh" }} />
          </Button>
          <Button
            style={buttonStyles}
            variant="danger"
            onClick={deleteMultipleTasks}
            disabled={isChangeStatusDisabled()}
            id="DeleteMultBtn"
          >
            <img src={trash} style={{ height: "3vh" }} />
          </Button>


// at bottom of file:

const buttonStyle = {
              height: "5vh",
              display: "flex",
              justifyContent: "center",
              alignItems: "center",
              width: "10vw",
            }

```

nadiacollado replied:
The style is the same for the buttons on lines 134 and 148:

```
            style={{
              height: "5vh",
              display: "flex",
              justifyContent: "center",
              alignItems: "center",
              width: "10vw",
            }}
```

If it's just one element then you don't necessarily have to extract the styling, but tbh I prefer a separate CSS file or at least styling that's removed from the return statement. I think it helps keep the component cleaner.

fravila08 commented:
I have renamed most tests to provide a better description for it's purpose, I've also moved any repeated styling out from the TSX files into a CSS file but have kept some in-line styling for some components with unique styling needs. 
The try and catch blocks for the front-end became necessary when working with integrations tests which is the reason for their existence, if the reviewers don't mind I would like to keep those for future implementation or growth in integration testing.

nadiacollado commented on to_do_client/src/App.css:
```diff
@@ -91,16 +91,24 @@ h1 {
   align-items: center;
 }
 
+.buttonHolder{
```
format spacing here :)

nadiacollado commented on to_do_client/src/App.css:
```diff
@@ -91,16 +91,24 @@ h1 {
+  display: flex;
+}
+
+.buttonImg{
```
and here as well

nadiacollado commented on to_do_client/src/App.tsx:
```diff
@@ -35,6 +35,10 @@ function App() {
     getResponse();
   }, []);
 
+  useEffect(()=>{
```
Do we need this useEffect/console?

nadiacollado commented on to_do_client/src/__tests__e2e_/ChangeTaskTitleE2E.test.tsx:
```diff
@@ -0,0 +1,48 @@
+    await browser.close();
+  });
+
+  describe("Change Task Title", () => {
```
I don't think you need this describe block as you already have it set up on line 4

nadiacollado commented on to_do_client/src/__tests__e2e_/DeletingATaskE2E.test.tsx:
```diff
@@ -14,9 +14,9 @@ describe("Deleteting a single task", () => {
-  afterAll(async () => {
-    await browser.close();
-  });
+  // afterAll(async () => {
```
delete commented out code
